feat#101/mysql 비동기 배치처리 작성

bellringstar opened the pull request:
## 📄 작업 설명
- QueueService를 중심으로 구매 요청을 처리하는 벌크 연산 시스템을 구축했습니다.
- 인메모리 큐를 구현하여 구매 요청을 버퍼링 하고 스케줄러를 통해 주기적으로 처리합니다.
- 프로세스의 흐름은 다음과 같습니다.
    - 결제 페이지 진입이 된 상황부터 시작됩니다.
    - 사용자가 결제 요청을 합니다.
    - 해당 티켓의 유효성을 검사하고 재고를 감소시킵니다.
    - 외부 결제 API로 요청을 합니다. 이는 비동기로 동작합니다.
    - 이 과정에서 바로 사용자에게는 paymentId가 리턴됩니다.
    - 그 후 스케줄러를 통해 주기적으로 해당 결제의 결과를 확인합니다.
    - 결제에 실패시 감소시켰던 재고를 원복시켜줍니다.
    - 결제에 성공시 InMemoryQueue에 정보를 추가합니다.
    - 그 후 스케줄러를 통해 InMemoryQueue에 있는 정보를 꺼내 배치처리를 합니다.
        - 배치처리 되는 내역은 구매내역, 체크인 내역입니다. 
## 🚨 관련 이슈
closes #101 

## 🌈 작업 상황
- 현재는 벌크연산의 부분허용 여부를 어떻게 할지 결정 못했습니다.
- 배치 작업 실패시 리트라이를 하는데 최종 실패시 영구저장소에 저장하고 복구하는 로직을 어떻게 할지 고민중입니다.
- 결제가 완료된 이후의 배치작업이기 때문에 서버가 다운되면 큐에 있는 내용이 다 유실됩니다. 따라서 대응책을 고민중입니다.
- 구매 가능 여부 검증은 총 2가지가 필요합니다.
    - 1인 1매 티켓 구매 제한 : 이 부분은 db에 유니크 제약 조건을 추가한다고 하여 애플리케이션에서는 생략했습니다.
    - 티켓 구매 가능 시간 판단에는 티켓 정보에 대한 캐싱을 도입하여 처리하였습니다. 
## 📌 기타

Commit:
feat: InMemoryQueue 구현을 위한 인터페이스 작성

- InMemoryQueue의 동작을 정의

Commit:
feat: InMemoryQueue 구현

- LinkedBlockingQueue를 사용해 구현
- 스레드 안정성과 블로킹 연산을 사용해 동시성 처리
- 큐가 가득 차거나 5초 동안 추가할 수 없는 경우 예외 발생
- 배치처리를 위한 메서드 추가

Commit:
feat: InMemoryQueue에서 발생할 커스텀 예외 발생

- 기존 ApiException과 달리 Api예외가 아니기 때문에 별도 커스텀 예외 생성

Commit:
feat: Queue에 추가할 dto 작성

Commit:
feat: InMemoryQueue 빈 등록

Commit:
feat: InMemoryQueue 모니터링을 위한 Indicator 추가

Commit:
refactor: 큐 사이즈를 상수로 관리하도록 변경

Commit:
refactor: InMemoryQueue 개선: 동시성 및 안정성 향상

- LinkedBlockingQueue를 ConcurrentLinkedQueue로 교체하여 락-프리 동작 구현
- AtomicInteger와 CAS를 사용한 정확한 크기 관리 구현
- 큐가 가득 찼을 때를 위한 지수 백오프 전략 추가
- offer와 poll 작업의 스레드 안전성 개선

Commit:
test: InMemoryQueueTest

- InMemoryQueue 동시성 테스트
- InMemoryQueue 동작 테스트

Commit:
feat: CustomQueueHealthIndicator.java

- 불필요 파일 삭제

Commit:
feat: InMemoryQueue 기본 생성자 추가

- 기본 큐 사이즈로 생성되는 기본 생성자 추가

Commit:
feat: QueueService 작성

- InMemoryQueue로 부터 데이터를 배치 사이즈 만큼 꺼내 db에 배치 인서트를 하는 서비스
- 동적으로 배치 사이즈를 결정하도록 처리
- 배치 처리에 실패한 내역들을 별도로 관리해 재시도 로직 추가

Commit:
feat: clear 추가

- 큐를 비우는 clear 메서드 추가

Commit:
docs: QueueService 동작 주석 추가

- 동작의 설명을 위한 주석 추가

Commit:
test: QueueServiceTest 코드 작성

- QueueService 동작 검증을 위한 테스트 코드

Commit:
feat: CacheConfig 정책 수정

- maximumSize 1000으로 수정

Commit:
feat: CachedTicketInfo 작성

- 캐싱을 위한 dto 작성

Commit:
feat: TicketCacheService 작성

- 티켓의 정보를 캐싱하는 서비스
- Caffeine을 사용해 구현

Commit:
feat: TicketCacheServiceTest 작성

- 티켓의 정보를 캐싱하는 서비스의 테스트

Commit:
refactor: 불필요 import 삭제

Commit:
feat: QueueService 개선

- 결제나 유효성 검증은 앞 단에서 하고 이 서비스는 배치 처리에만 집중하도록 변경

Commit:
test: QueueService 검증을 위한 테스트코드

Commit:
feat: 결제시스템의 리스폰스 dto 생성

- 유일성이 있는 구매 id를 리턴

Commit:
feat: 티켓 정보 캐싱에 festivalId 추가

Commit:
feat: PaymentService 구현

- CompletableFuture를 이용해 비동기적으로 처리
- paymentId를 통해 상태 추적에 용이하도록 처리
- Caffeine을 사용해 결제 상태 캐싱

Commit:
feat: PaymentService 구현

- CompletableFuture를 이용해 비동기적으로 처리
- paymentId를 통해 상태 추적에 용이하도록 처리
- Caffeine을 사용해 결제 상태 캐싱

Commit:
feat: PaymentService 구현

- CompletableFuture를 이용해 비동기적으로 처리
- paymentId를 통해 상태 추적에 용이하도록 처리
- Caffeine을 사용해 결제 상태 캐싱

Commit:
feat: PurchaseFacadeService 구현 변경

- 구매 요청을 검증
- 외부 결제 시스템에 결제를 요청
- 결제 상태를 스케줄링을 통해 관리하고 결과에 따른 작업처리

Commit:
feat: PurchaseController 수정

- 결제 api에서 paymentId를 리턴해 상태를 추적할 수 있도록 변경
- 결제 상태를 확인할 수 있는 api 작성

Commit:
feat: PurchaseController에서 결제 상태 전달을 위한 response dto

Commit:
test: PurchaseFacadeService 변경에 따른 테스트코드 수정

Commit:
feat: 하위호환을 위해 코드 복

Commit:
feat: PurchaseControllerV2 작성

- 하위호환성을 위해 기존 코드를 놔두고 v2로 작성

Commit:
test: PurchaseControllerV2Test 작성

Commit:
feat: 새로운 메서드 v2로 이동

Commit:
refactor: 모듈화로 인한 패키지 위치 변경

Commit:
docs: PurchaseControllerV2 문서화

minji-park-eng commented on backend/api-server/src/main/java/com/wootecam/festivals/domain/payment/service/PaymentService.java:
```diff
@@ -0,0 +1,109 @@
+        // 비동기적으로 결제 처리 시작
+        // CompletableFuture.runAsync()는 별도의 스레드에서 작업을 실행합니다.
+        // 이 메서드는 즉시 반환되며, 결제 처리는 백그라운드에서 계속됩니다.
+        CompletableFuture.runAsync(() -> processPayment(paymentId, memberId, ticketId));
```
p4: CompletableFuture가 완료되면 이벤트를 발행해서 결제 완료 배치 처리 큐에 넣으면 어떨까요?

bellringstar replied:
현재는 스케쥴기반이지만 추후 그렇게 변경하겠습니다.!

kkyu0718 reviewed (approved):
수고많으셨습니다

HyeonJun0530 committed:
refactor#115/spring 성능 개선 (#125)

* refactor: eager -> lazy 로딩으로 변경

* fix: @EnableAsync 누락에 따른 어노테이션 추가

* chore: 커넥션 풀, 쓰레드 풀 설정 및 osiv Off

* refactor: 필요없는 검증 로직 삭제

- 페스티벌 조회 후 검증하는 로직 삭제

* refactor: pay 로직 non-blocking 처리

- 기존 Thread.sleep blocking에서 non-blocking으로 수정

* refactor: 체크인 중복 로직에서 기존 체크인 엔티티 -> id만 가져오도록 변경

* refactor: 세션 확인에서 memberRepository 제외

- 세션에서 값을 가져와 null 체킹만 하도록 변경
- 세션이니까 가져와서 서비스에서 FindBy해서 어차피 검증하니까 세션만 있으면 Authentication에 넣어주고 여기서 세션에서 가져온 memberId만 사용해서 실제 member 조회는 각 서비스 조회해서 사용하는 것이 좋겠다 !

* config: osiv off 및 커넥션 풀, 쓰레드 풀 옵션 적용

- hibernate 정보 수집을 위한 로그 추가
- 톰캣, 스프링 쓰레드 풀 옵션 추가
- 커넥션 풀 옵션 추가
- osiv off

* refactor: fk 부분 인덱스 추가 및 sql ddl 변경

* test: TicketService, Authentication 로직 변경에 따른 테스트 수정

---------

Co-authored-by: HyeonJun0530 <hyjn053@gmail.com>

HyeonJun0530 committed:
refactor#115/spring 성능 개선 (#125)

* refactor: eager -> lazy 로딩으로 변경

* fix: @EnableAsync 누락에 따른 어노테이션 추가

* chore: 커넥션 풀, 쓰레드 풀 설정 및 osiv Off

* refactor: 필요없는 검증 로직 삭제

- 페스티벌 조회 후 검증하는 로직 삭제

* refactor: pay 로직 non-blocking 처리

- 기존 Thread.sleep blocking에서 non-blocking으로 수정

* refactor: 체크인 중복 로직에서 기존 체크인 엔티티 -> id만 가져오도록 변경

* refactor: 세션 확인에서 memberRepository 제외

- 세션에서 값을 가져와 null 체킹만 하도록 변경
- 세션이니까 가져와서 서비스에서 FindBy해서 어차피 검증하니까 세션만 있으면 Authentication에 넣어주고 여기서 세션에서 가져온 memberId만 사용해서 실제 member 조회는 각 서비스 조회해서 사용하는 것이 좋겠다 !

* config: osiv off 및 커넥션 풀, 쓰레드 풀 옵션 적용

- hibernate 정보 수집을 위한 로그 추가
- 톰캣, 스프링 쓰레드 풀 옵션 추가
- 커넥션 풀 옵션 추가
- osiv off

* refactor: fk 부분 인덱스 추가 및 sql ddl 변경

* test: TicketService, Authentication 로직 변경에 따른 테스트 수정

---------

Co-authored-by: HyeonJun0530 <hyjn053@gmail.com>

Commit:
refactor: purchase 컨트롤러 통합

- V2와 기존 컨트롤러 통합

Commit:
refactor: purchase 컨트롤러 통합에 따른 문서, 테스트 수정

Commit:
feat: 재고 차감 위치 변경

- 결제 로직에서 재고를 차감하지 않고 페이지에 접근하기 전 결제 페이지 접근 권한을 주며 재고를 차감
- 결제에 실패시 재고를 원복해주도록 처

Commit:
fix: InMemoryQueue clear 메서드 수정

- 기존 size를 초기화해주지 않던 버그 수정

Commit:
test: InMemoryQueueTest

- 클래스 기본 동작 테스트 추가

Commit:
feat: Checkin null 체크 메시지 추

Commit:
test: InMemoryQueueTest 추

Commit:
test: PaymentServiceTest 작성

Commit:
feat: purchaseData null 체크 추가

Commit:
test: QueueServiceTest 테스트 추가

Commit:
test: InMemoryQueueTest 테스트 수

Commit:
test: QueueServiceIntegrationTest 테스트 수정

Commit:
refactor: PaymentService 불필요 로직 삭

minji-park-eng reviewed (commented):
고생하셨습니다!! :)

Commit:
feat: 결제 결과 시뮬레이션에서 PENDING 제거

Commit:
refactor: 불필요 의존성 삭제

Commit:
fix: 배치처리 삽입 칼럼 수정

- created_at, updated_at 처리

Commit:
test: 큐 사이즈 테스트 수

Commit:
feat: 세션에 저장되는 값 변

Commit:
test: 티켓 구매 로직 변경에 따른 테스트코드 수

Commit:
feat: 티켓 재고 감소에서 멤버id 삽입하기로 정책 변경에 따른 코드 수정

- 원복시 재고를 +1 해주는게 아닌 멤버id를 null로 변경하는 방법으로 변경

Commit:
test: 결제 로직 변경에 따른 테스트 수정

Commit:
fix: ticketStockId를 전달받아야 하는 곳에서 ticketId를 전달받던 문제 수정

Commit:
refactor: 매개변수명 변경

Commit:
test: PurchaseData 필드 추가에 따른 테스트 코드 수
